feat: show color swatches in case table

kswenson opened the pull request.

tealefristoe commented on v3/src/global.d.ts:
```diff
@@ -4,6 +4,14 @@ declare module "*.png";
 
+declare module "color-parse" {
+  export default function parse(cstr: string): {
+    space: string | undefined
```
```suggestion
    space?: string
```

tealefristoe commented on v3/src/global.d.ts:
```diff
@@ -4,6 +4,14 @@ declare module "*.png";
 declare module "d3-v6-tip";
 
+declare module "color-parse" {
+  export default function parse(cstr: string): {
```
Any way to use the `ParseColorResult` interface defined in `color-parse-strict.ts` for the return type here?

tealefristoe reviewed (approved):
Looks good 👍 

I left a couple of minor comments. I also notice that there are some stylistic differences between `color-parse-strict.ts` and the rest of the code base. I'm not sure if these are worth addressing since it's borrowed code, but I'd consider at least replacing `var` with `let`.

tealefristoe commented:
I also just noticed that the `attribute-types.spec.ts` cypress test is failing.

bfinzer commented:
Bug: I made a new table, renamed the attribute to `color` and typed "red" into the first cell. The red color swatch appeared. It shouldn't. When I added more cases, that cell value reverted to string "red" as it should be.

kswenson commented:
>I made a new table, renamed the attribute to `color` and typed "red" into the first cell. The red color swatch appeared. It shouldn't.

That's bizarre -- that was one of my test cases and it was working for me. 🤔 I'll take a look.

bfinzer commented on v3/src/components/case-table/use-columns.tsx:
```diff
@@ -25,6 +27,28 @@
+  const type = attr?.type
+
+  // colors
+  if (type === "color" && parseColor(str).space) {
```
Looks like if the type is not "color" the swatch will never show. But we want "strict" colors to display as swatches always, right?

kswenson replied:
That wasn't my understanding from our previous conversation, but we can discuss at the planning meeting today.

bfinzer commented on v3/src/utilities/color-parse-strict.ts:
```diff
@@ -0,0 +1,169 @@
+	let m: RegExpExecArray | null
+	var parts: Array<string | number> = [], alpha = 1
+	let space = ''
+
```
Get rid of commented out code below?

bfinzer reviewed (changes requested):
I noted a number of bugs. Some (or all?) are case table refresh bugs and could be set aside in PT as you see fit.

kswenson commented:
> I also just noticed that the `attribute-types.spec.ts` cypress test is failing.

I fixed the cypress error which was preventing deployment, which may also explain some of the bugs @bfinzer found.

>I made a new table, renamed the attribute to `color` and typed "red" into the first cell. The red color swatch appeared. It shouldn't.

@bfinzer I also fixed one internal invalidation bug that explained some of this, but I think some of the other bugs you found were the result of the previous build not deploying due to the failing cypress test, so you may want to test again.

bfinzer commented:
I checked out the branch directly. Wouldn’t that mean deployment is not an issue?

> On Mar 6, 2024, at 11:56 AM, Kirk Swenson ***@***.***> wrote:
> 
> 
> I also just noticed that the attribute-types.spec.ts cypress test is failing.
> 
> I fixed the cypress error which was preventing deployment, which may also explain some of the bugs @bfinzer <https://github.com/bfinzer> found.
> 
> I made a new table, renamed the attribute to color and typed "red" into the first cell. The red color swatch appeared. It shouldn't.
> 
> @bfinzer <https://github.com/bfinzer> I also fixed one internal invalidation bug that explained some of this, but I think some of the other bugs you found were the result of the previous build not deploying due to the failing cypress test, so you may want to test again.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/concord-consortium/codap/pull/1148#issuecomment-1981678674>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABLKA72NSUDKVG2V2UJZJP3YW5YGXAVCNFSM6AAAAABEHT6VJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBRGY3TQNRXGQ>.
> You are receiving this because you were mentioned.
>

bfinzer commented:
I notice the branch name is missing the `v3` suffix.

bfinzer reviewed (approved):
👍🏻LGTM
The bugs I noted are no longer present. Nice!
